upstairs/src/client.rs

```diff
@@ -892,57 +878,62 @@ impl DownstairsClient {
         self.state = DsState::Active;
     }
 
+    /// Checks whether the given job should be sent
+    ///
+    /// Returns `true` if it should be sent and `false` otherwise
+    ///
+    /// If the job should be skipped, then it is added to `self.skipped_jobs`.
+    /// `self.io_state_count` is updated with the incoming job state.
+    #[must_use]
     pub(crate) fn enqueue(
         &mut self,
-        io: &mut DownstairsIO,
+        ds_id: JobId,
+        io: &IOop,
         last_repair_extent: Option<ExtentId>,
-    ) -> IOState {
-        assert_eq!(io.state[self.client_id], IOState::New);
-
+    ) -> bool {
         // If a downstairs is faulted or ready for repair, we can move
         // that job directly to IOState::Skipped
         // If a downstairs is in repair, then we need to see if this
         // IO is on a repaired extent or not.  If an IO spans extents
         // where some are repaired and some are not, then this IO had
         // better have the dependencies already set to reflect the
         // requirement that a repair IO will need to finish first.
-        let r = match self.state {
+        let should_send = match self.state {
+            // We never send jobs if we're in certain inactive states
             DsState::Faulted
             | DsState::Replaced
             | DsState::Replacing
-            | DsState::LiveRepairReady => {
-                io.state.insert(self.client_id, IOState::Skipped);
-                self.skipped_jobs.insert(io.ds_id);
-                IOState::Skipped
-            }
+            | DsState::LiveRepairReady => false,
+
+            // We conditionally send jobs if we're in live-repair, depending on
+            // the current extent.
             DsState::LiveRepair => {
                 // Pick the latest repair limit that's relevant for this
                 // downstairs.  This is either the extent under repair (if
                 // there are no reserved repair jobs), or the last extent
                 // for which we have reserved a repair job ID; either way, the
                 // caller has provided it to us.
-                if io.work.send_io_live_repair(last_repair_extent) {
-                    // Leave this IO as New, the downstairs will receive it.
-                    IOState::New
-                } else {
-                    // Move this IO to skipped, we are not ready for
-                    // the downstairs to receive it.
-                    io.state.insert(self.client_id, IOState::Skipped);
-                    self.skipped_jobs.insert(io.ds_id);
-                    IOState::Skipped
-                }
+                io.send_io_live_repair(last_repair_extent)
             }
-            _ => IOState::New,
+
+            // Otherwise, we always send jobs!
+            //
+            // XXX should we be pickier here? e.g. if we're somehow in
```

Comment:
Yes, we should be pickier.
Really, there should be no `_ => ` catch all I believe.

Was there any checking in `in_progress` for other downstairs states?
It's also possible that we were not checking and also never created a situation
where a downstairs was in `DsState::BadVersion`, so we never hit this.